### PR TITLE
Support filters in wrench

### PR DESCRIPTION
**Pull request by rlhunt**
@jrmuizel

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/801)

### CODE REVIEW

**Comment by glennw** [comment]
The reftest is failing on CI - I haven't looked into why yet.

**Comment by rlhunt** [comment]
Okay, I got run_tests.py/headless working locally on my machine.

The reftest failure is in the opacity test. Using 50% opacity on rgb(0,255,0) is giving rgb(127,255,127), while I hardcoded the reference to rgb(128,255,128). It was passing on my machine so I'm guessing it's hardware/software related. I don't know of any easy fix besides adding reftest fuzzing, so I'll remove the reftest for now.

**Comment by glennw** [comment]
The tests are always run on the OSMesa software rasterizer, and we compile a specific version of that to be used on the CI machines. So it should be fine to just update the reftest to expect the value that you're seeing on the CI machines.

**Comment by glennw** [comment]
(Or to be more specific, when running the tests locally, we should be using the headless mode - this will mean tests run locally *should* always match the results on CI).

File: wrench/src/yaml_helper.rs [hunk 1]
```diff
@@ -319,4 +321,53 @@ impl YamlHelper for Yaml {
     fn as_scroll_policy(&self) -> Option<ScrollPolicy> {
         return self.as_str().and_then(|string| string_to_scroll_policy(string))
     }
+
+    fn as_filter_op(&self) -> Option<FilterOp> {
+        if let Some(s) = self.as_str() {
```
Comment (kvark) [inline wrench/src/yaml_helper.rs]: this function body is overly verbose and could be reduced by some refactoring
not required for this PR though

File: wrench/src/yaml_frame_reader.rs [hunk 2]
```diff
@@ -492,8 +492,7 @@ impl YamlFrameReader {
             }
         }
 
-        // FIXME handle these
-        let filters: Vec<FilterOp> = Vec::new();
+        let filters: Vec<FilterOp> = yaml["filters"].as_vec_filter_op().unwrap_or(vec![]);
```
Comment (kvark) [inline wrench/src/yaml_frame_reader.rs]: nit: doesn't need explicit type here

File: wrench/src/yaml_helper.rs [hunk 3]
```diff
@@ -319,4 +321,53 @@ impl YamlHelper for Yaml {
...
+        } else {
+            None
+        }
+    }
+
+    fn as_vec_filter_op(&self) -> Option<Vec<FilterOp>> {
+        if let Some(v) = self.as_vec() {
+            Some(v.iter().map(|x| x.as_filter_op().unwrap()).collect())
+        } else {
+            if let Some(op) = self.as_filter_op() {
```
Comment (kvark) [inline wrench/src/yaml_helper.rs]: this asks for `self.as_filter_op().map(|op| vec![op])`

File: wrench/src/yaml_helper.rs (hunk 1 above)
Comment (rlhunt) [inline wrench/src/yaml_helper.rs]: Yeah I definitely agree. It'd be nice to have something like define-enum-conversion that works with more general enums.

**Comment by rlhunt** [comment]
Investigating the opacity reftest further, it looks like it's not just the color that's off but there is a one pixel border on the bottom and right edges that is a different color than the rest of the rect.

For now I think it's easiest to remove the test and maybe add it in later.

**Pull request merged**